tests/test_caching.py

```diff
@@ -0,0 +1,580 @@
+from devito.types.basic import Scalar, Symbol
+
+pytestmark = skipif(['yask', 'ops'])
+
+
+@pytest.fixture
+def operate_on_empty_cache():
+    """
+    To be used by tests that assert against the cache size. There are two
+    reasons this is required:
+
+        * Most symbolic objects embed further symbolic objects. For example,
+          Function embeds Dimension, DerivedDimension embed a parent Dimension,
+          and so on. The embedded objects require more than one call to
+          `clear_cache` to be evicted (typically two -- the first call
+          evicts the main object, then the children become unreferenced and so
+          they are evicted upon the second call). So, depending on what tests
+          were executed before, it is possible that one `clear_cache()` evicts
+          more than expected, making it impossible to assert against cache sizes.
+        * Due to some global symbols in `conftest.py`, it is possible that when
+          for example a SparseFunction is instantiated, fewer symbolic object than
+          expected are created, since some of them are available from the cache
+          already.
+    """
+    old_cache = _SymbolCache.copy()
+    _SymbolCache.clear()
+    yield
+    _SymbolCache.update(old_cache)
+
+
+class TestHashing(object):
+
+    """
+    Test hashing of symbolic objects.
+    """
+
+    def test_constant(self):
+        """Test that different Constants have different hash value."""
+        c0 = Constant(name='c')
+        c1 = Constant(name='c')
+        assert c0 is not c1
+        assert hash(c0) != hash(c1)
+
+    def test_dimension(self):
+        """Test that different Dimensions have different hash value."""
+        d0 = Dimension(name='d')
+        s0 = Scalar(name='s')
+        d1 = Dimension(name='d', spacing=s0)
+        assert hash(d0) != hash(d1)
+
+        s1 = Scalar(name='s', dtype=np.int32)
+        d2 = Dimension(name='d', spacing=s1)
+        assert hash(d1) != hash(d2)
+
+        d3 = Dimension(name='d', spacing=Constant(name='s1'))
+        assert hash(d3) != hash(d0)
+        assert hash(d3) != hash(d1)
+
+    def test_sub_dimension(self):
+        """Test that different SubDimensions have different hash value."""
+        d0 = Dimension(name='d')
+        d1 = Dimension(name='d', spacing=Scalar(name='s'))
+
+        di0 = SubDimension.middle('di', d0, 1, 1)
+        di1 = SubDimension.middle('di', d1, 1, 1)
+        assert hash(di0) != hash(d0)
+        assert hash(di0) != hash(di1)
+
+        dl0 = SubDimension.left('dl', d0, 2)
+        assert hash(dl0) != hash(di0)
+
+    def test_conditional_dimension(self):
+        """Test that different ConditionalDimensions have different hash value."""
+        d0 = Dimension(name='d')
+        s0 = Scalar(name='s')
+        d1 = Dimension(name='d', spacing=s0)
+
+        cd0 = ConditionalDimension(name='cd', parent=d0, factor=4)
+        cd1 = ConditionalDimension(name='cd', parent=d0, factor=5)
```

Comment:
if factor=4 and yet indirect=False, asserts in lines 89 and 90 hold ?

Reply:
no becasue indirect's default value is False